File: 2610.java
import java.util.Scanner;


public class Main{
 public static void main(String[] args)
 {
  Scanner in = new Scanner(System.in);
  double x1 = in.nextDouble();
  double y1 = in.nextDouble();
  double x2 = in.nextDouble();
  double y2 = in.nextDouble();
  //double r1 = 0;
  //double r2 = 0;
  //double d3 = 0;
  boolean canEscape = false;
  while(in.hasNext())
   {
	String s1 = in.next();
	in.skip(" ");
	String s2 = in.nextLine();
	//in.nextLine();
	double x3 = Double.parseDouble(s1);
	double y3 = Double.parseDouble(s2);
	double d1 = (x3 - x1)*(x3 - x1) + (y3 - y1)*(y3 - y1);
	double d2 = (x3 - x2)*(x3 - x2) + (y3 - y2)*(y3 - y2);
	if(d2 > 4 *  d1)
	{
	  canEscape = true;
	  System.out.print("The gopher can escape through the hole at ("+s1+','+s2+").");
	   break;
	}
    }

    if(canEscape == false)
	System.out.println("The gopher cannot escape.");
  }
}

File: 2610.cc
#include <iostream>
#include <cstdlib>
#include <cstring>
#include <cstdio>
#include <cmath>
using namespace std;

#define eps 1.0E-8

double xx1, xx2, yy1, yy2;

double dist(double x1, double y1, double x2, double y2) {
  double x = abs(x1 - x2);
  double y = abs(y1 - y2);
  return sqrt(x * x + y * y);
}

bool ok(double x, double y) {
  double d1 = dist(xx1, yy1, x, y);
  double d2 = dist(xx2, yy2, x, y);
  if (d1 * 2 + eps < d2) return true;
  return false;
}

int main() {
  scanf("%lf%lf%lf%lf", &xx1, &yy1, &xx2, &yy2);
  double x, y;
  while (scanf("%lf%lf", &x, &y) != EOF) {
    if (ok(x, y)) {
      printf("The gopher can escape through the hole at (%.3f,%.3f).\n", x, y);
      return 0;
    }
  }
  printf("The gopher cannot escape.\n");
  return 0;
}
